lib/core/include/data_object_proxy.hpp

```diff
@@ -0,0 +1,212 @@
+        /// \since 4.2.9
+        auto data_id() const -> rodsLong_t { return doi_->dataId; }
+
+        /// \returns rodsLong_t
+        /// \retval id of the collection containing this data object
+        /// \since 4.2.9
+        auto collection_id() const -> rodsLong_t { return doi_->collId; }
+
+        /// \returns std::string_view
+        /// \retval Logical path of the data object
+        /// \since 4.2.9
+        auto logical_path() const -> std::string_view { return doi_->objPath; }
+
+        /// \returns std::string_view
+        /// \retval Owner user name of the data object
+        /// \since 4.2.9
+        auto owner_user_name() const -> std::string_view { return doi_->dataOwnerName; }
+
+        /// \returns std::string_view
+        /// \retval Owner zone name of the data object
+        /// \since 4.2.9
+        auto owner_user_zone() const -> std::string_view { return doi_->dataOwnerZone; }
+
+        /// \returns size_type
+        /// \retval Number of replicas for this data object
+        /// \since 4.2.9
+        auto replica_count() const -> size_type { return r_.size(); }
+
+        /// \returns const replica_list&
+        /// \retval Reference to the list of replica_proxy objects
+        /// \since 4.2.9
+        auto replicas() const -> const replica_list& { return r_; }
+
+        /// \returns const doi_pointer_type
+        /// \retval Pointer to the underlying struct
+        /// \since 4.2.9
+        auto get() const -> const doi_pointer_type { return doi_; }
+
+        // mutators
+
+        /// \brief Set the data id for the data object
+        /// \param[in] _did - New value for data id
+        /// \returns void
+        /// \since 4.2.9
+        template<
+            typename P = doi_type,
+            typename = std::enable_if_t<!std::is_const_v<P>>>
+        auto data_id(const rodsLong_t _did) -> void
+        {
+            for (auto& r : r_) {
+                r.data_id(_did);
+            }
+        }
+
+        /// \brief Set the collection id for the data object (i.e. all replicas)
+        /// \param[in] _cid - New value for collection id
+        /// \returns void
+        /// \since 4.2.9
+        template<
+            typename P = doi_type,
+            typename = std::enable_if_t<!std::is_const_v<P>>>
+        auto collection_id(const rodsLong_t _cid) -> void 
+        {
+            // TODO: should this set the logical path too?
+            for (auto& r : r_) {
+                r.collection_id(_cid);
+            }
+        }
+
+        /// \brief Set the logical path for the data object (i.e. all replicas)
+        /// \param[in] _lp - New value for logical path
+        /// \returns void
+        /// \since 4.2.9
+        template<
+            typename P = doi_type,
+            typename = std::enable_if_t<!std::is_const_v<P>>>
+        auto logical_path(std::string_view _lp) -> void
+        {
+            // TODO: should this set the collId too?
```

Comment:
i think it must?    what does it mean not to?

Reply:
See comment in collection_id above